### DISCUSSION TITLE
请问如何区分 bot.send发送的消息和从其它客户端同步过来的消息？

### DISCUSSION THREAD

**Discussion by Misaka-Mikoto-Tech** [nonebot/discussions#338] [问答]
因为要把bot作为正常的QQ使用，因此会在其它客户端发消息，但bot自身也在send消息，但bot不能处理自己send的消息，否则会陷入自问自答的循环，但是 on_message 收到的MessageEvent.post_type 都是 message_sent, 其它字段也无法区分，所以有什么办法可以区分这两种消息呢？

**Comment by Misaka-Mikoto-Tech** [comment]
已通过钩子函数 @Bot.on_calling_api 搞定，虽然不太优雅。另外我发现 bot.send_group_msg 返回的 message_id 都是正数，而从其它设备同步的 message_sent 消息的 message_id 都是负数，不知道这是否是官方的规则，如果是官方的规则那可以用正负号判断，否则只能自己记录了

**Comment by yanyongyu** [comment]
其他客户端发送的消息post type是message_sent

**Reply by Misaka-Mikoto-Tech** [reply]
> 其他客户端发送的消息post type是message_sent

自己发送的消息也是 message_sent

**Reply by yanyongyu** [reply]
那这个没法区分了（

**Comment by YangRucheng** [comment]
如果是go-cqhttp, 好像可以选择是否上报自身发出的消息
```
  # 是否上报自身消息
  report-self-message: false
```
如果你用钩子函数, 为什么不在事件预处理时打断 `self_id` 和 `user_id` 相同的消息事件呢?
